src/cressi_goa_parser.c

```diff
@@ -271,85 +620,177 @@ cressi_goa_parser_get_field (dc_parser_t *abstract, dc_field_type_t type, unsign
+		unsigned int sample_rate = header[10];
+		/* Valid sample rates are: 0.5s, 1s, 2s, 5s */
+		switch (sample_rate) {
+			case 1:
+				interval = 1;
+				break;
+			case 2:
+				interval = 2;
+				break;
+			case 3:
+				interval = 4;
+				break;
+			case 4:
+				interval = 10;
+				break;
+			default:
+				ERROR (abstract->context, "Unknown sample rate: 0x%02x.", sample_rate);
+				return DC_STATUS_DATAFORMAT;
+		}
 
-	unsigned int divemode = data[2];
-	if (divemode >= C_ARRAY_SIZE(layouts)) {
-		return DC_STATUS_DATAFORMAT;
+		/* Adjust the size in order to skip the Advanced Freedive Dip Stats that come at the end.
+		 * If we'd ever want to use those, the structure is:
+		 *
+		 * SURFTIME | MAXDEPTH | DIPTIME | MINTEMP | SPEED_DESC | SPEED_ASC | TARAVANA_VIOLATION
+		 *   u16    |   u16    |   u16   |   u16   |    u16     |    u16    |       bool
+		 */
+		unsigned int num_samples = array_uint16_le (header + 2);
+		size = offset + num_samples * 2;
 	}
 
-	const cressi_goa_layout_t *layout = &layouts[divemode];
-
-	if (size < layout->headersize)
-		return DC_STATUS_DATAFORMAT;
-
-	unsigned int interval = divemode == FREEDIVE ? 2 : 5;
-
 	unsigned int time = 0;
 	unsigned int depth = 0;
 	unsigned int gasmix = 0, gasmix_previous = 0xFFFFFFFF;
 	unsigned int temperature = 0;
-	unsigned int have_temperature = 0;
-	unsigned int complete = 0;
+	bool have_temperature = false;
+	bool complete = false;
 
-	unsigned int offset = layout->headersize;
 	while (offset + 2 <= size) {
 		dc_sample_value_t sample = {0};
 
 		// Get the sample type and value.
 		unsigned int raw = array_uint16_le (data + offset);
 		unsigned int type  = (raw & 0x0003);
 		unsigned int value = (raw & 0xFFFC) >> 2;
+		unsigned int t = 0;
 
-		if (type == DEPTH || type == DEPTH2) {
-			depth =  (value & 0x07FF);
-			gasmix = (value & 0x0800) >> 11;
+		if (type == DEPTH_SCUBA) {
+			depth = value & 0x7ffu;
+			gasmix = (value >> 11) & ((layout->version < 3) ? 0x1u : 0x3u);
+			/* speed_level = (layout->version < 3) ? (raw >> 14) & 0x3u : 0; */
 			time += interval;
-			complete = 1;
+			complete = true;
+		} else if (type == DEPTH_FREE) {
+			depth = value & ((layout->version < 4) ? 0x7ffu : 0xfffu);
+			time += interval;
+			complete = true;
 		} else if (type == TEMPERATURE) {
 			temperature = value;
-			have_temperature = 1;
-		} else if (type == TIME) {
-			time += value;
+			have_temperature = true;
+		} else if (type == SURFACE) {
+			// SURFACE values are not given in the sample rate, but as seconds
+			if (divemode == FREEDIVE_ADV && (offset + 2 < size)) {
+				if (time)
+					t = interval;
+				if (SUBSURFACE_CORRECTION(interval, (time + t)))
+					t = 2;
+				time += t;
+				depth = 0;
+
+				// Time (seconds).
+				sample.time = time * cressi_goa_time_multiplier;
+				cressi_goa_parser_callback (DC_SAMPLE_TIME, &sample, &ctx);
+				// Depth (1/10 m).
+				sample.depth = 0.0;
+				cressi_goa_parser_callback (DC_SAMPLE_DEPTH, &sample, &ctx);
+			}
```

Comment:
I'm not sure whether those corrections are valid at all.

If we're fine with them happening I guess they could also be applied for all the other dive modes as well and not only for `FREEDIVE_ADV`.